src/nn/tensor_collection/collection.rs

```diff
@@ -0,0 +1,96 @@
+#![allow(clippy::type_complexity)]
+
+use crate::{
+    shapes::{Dtype, Shape},
+    tensor::{DeviceStorage, OneFillStorage, Tensor, ZeroFillStorage},
+};
+
+/// A collection of named tensors. Implementing this trait will enable anything
+/// that operates on tensors, like resetting, EMA, counting number of params,
+/// gradient updates, etc.
+pub trait TensorCollection<E: Dtype, D: DeviceStorage>: Sized {
+    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err>;
+}
+
+/// An object that can visit [TensorCollection]s and [Tensor]s recursively.
+pub trait ModuleVisitor<T, E: Dtype, D: DeviceStorage>: Sized {
+    type Err;
+
+    /// Visit a [TensorCollection]
+    fn visit_module<Field, GetRef, GetMut>(
+        &mut self,
+        name: &str,
+        get_refs: GetRef,
+        get_muts: GetMut,
+    ) -> Result<(), Self::Err>
+    where
+        GetRef: FnMut(&T) -> &Field,
+        GetMut: FnMut(&mut T) -> &mut Field,
+        Field: TensorCollection<E, D>;
+
+    /// Visits an actual named [Tensor]
+    fn visit_tensor<S: Shape, GetRef, GetMut>(
+        &mut self,
+        name: &str,
+        get_refs: GetRef,
+        get_muts: GetMut,
+        opts: TensorOptions<S, E, D>,
+    ) -> Result<(), Self::Err>
+    where
+        GetRef: FnMut(&T) -> &Tensor<S, E, D>,
+        GetMut: FnMut(&mut T) -> &mut Tensor<S, E, D>;
+}
+
+impl<S: Shape, E: Dtype, D: DeviceStorage> TensorCollection<E, D> for Tensor<S, E, D> {
+    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err> {
+        visitor.visit_tensor(
+            "",
+            |s| s,
+            |s| s,
+            TensorOptions {
+                do_gradient_update: true,
+                reset: |_| Ok(()),
+            },
+        )
+    }
+}
+
+/// Options to change behavior of [TensorVisitor]
+pub struct TensorOptions<S: Shape, E: Dtype, D: DeviceStorage> {
+    /// Whether the tensor should be updated with gradients
+    pub do_gradient_update: bool,
+    pub reset: fn(&mut Tensor<S, E, D>) -> Result<(), D::Err>,
+}
```

Comment:
I think we should add `#[non_exhaustive]` to this struct so that we can easily add new options in the future if needed. We could also define default values for all fields and use the builder pattern to construct this, but I'm not sure this is better.

Reply:
Oh just kidding, anyone implementing Visitors will need to access those fields. I'll do #[non_exhaustive]!